File: objectivec/day18_part2_2016.m
#import <Foundation/Foundation.h>

const int totalRows = 400000;

int countSafeTiles(NSString *firstRow);
BOOL isTrap(int left, int center, int right, NSString *row);
unichar safeIfOutOfBounds(int index, NSString *row);
int countChar(NSString *str, unichar character);

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSString *firstRow = [NSString stringWithContentsOfFile:@"input.txt" encoding:NSUTF8StringEncoding error:nil];
        int safeTilesCount = countSafeTiles(firstRow);
        NSLog(@"%d", safeTilesCount);
    }
    return 0;
}

int countSafeTiles(NSString *firstRow) {
    NSString *currentRow = firstRow;
    int safeCount = countChar(currentRow, '.');

    for (int i = 1; i < totalRows; i++) {
        NSMutableString *nextRow = [NSMutableString stringWithCapacity:currentRow.length];
        for (int j = 0; j < currentRow.length; j++) {
            if (isTrap(j - 1, j, j + 1, currentRow)) {
                [nextRow appendString:@"^"];
            } else {
                [nextRow appendString:@"."];
                safeCount++;
            }
        }
        currentRow = nextRow;
    }
    return safeCount;
}

BOOL isTrap(int left, int center, int right, NSString *row) {
    unichar l = safeIfOutOfBounds(left, row);
    unichar c = [row characterAtIndex:center];
    unichar r = safeIfOutOfBounds(right, row);

    return (l == '^' && c == '^' && r == '.') ||
           (c == '^' && r == '^' && l == '.') ||
           (l == '^' && c == '.' && r == '.') ||
           (r == '^' && c == '.' && l == '.');
}

unichar safeIfOutOfBounds(int index, NSString *row) {
    if (index < 0 || index >= row.length) {
        return '.';
    }
    return [row characterAtIndex:index];
}

int countChar(NSString *str, unichar character) {
    int count = 0;
    for (NSUInteger i = 0; i < str.length; i++) {
        if ([str characterAtIndex:i] == character) {
            count++;
        }
    }
    return count;
}

File: training_data/reflection-tuning/day18_part2_2016.go
package main

import (
	"fmt"
	"io/ioutil"
	"strings"
)

func countSafeTiles(initialRow string, totalRows int) int {
	safeTiles := strings.Count(initialRow, ".")
	currentRow := initialRow

	for i := 1; i < totalRows; i++ {
		nextRow := make([]byte, len(currentRow))
		for j := 0; j < len(currentRow); j++ {
			left := byte('.')
			if j > 0 {
				left = currentRow[j-1]
			}
			right := byte('.')
			if j < len(currentRow)-1 {
				right = currentRow[j+1]
			}
			center := currentRow[j]

			if (left == '^' && center == '^' && right == '.') ||
				(left == '.' && center == '^' && right == '^') ||
				(left == '^' && center == '.' && right == '.') ||
				(left == '.' && center == '.' && right == '^') {
				nextRow[j] = '^'
			} else {
				nextRow[j] = '.'
				safeTiles++
			}
		}
		currentRow = string(nextRow)
	}

	return safeTiles
}

func main() {
	input, err := ioutil.ReadFile("input.txt")
	if err != nil {
		panic(err)
	}
	initialRow := strings.TrimSpace(string(input))

	part1 := countSafeTiles(initialRow, 40)
	fmt.Println("Part 1:", part1)

	part2 := countSafeTiles(initialRow, 400000)
	fmt.Println("Part 2:", part2)
}
